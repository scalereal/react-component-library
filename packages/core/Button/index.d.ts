import { FC } from 'react';
import { WithStyle } from '../../utils';
import { Props } from './types';
export declare type Ref = HTMLButtonElement;
declare const Button: FC<Props> & WithStyle;
export default Button;
