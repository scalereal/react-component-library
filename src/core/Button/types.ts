import { ElementType, ReactNode } from 'react';
import { HTMLProps, Omit, WithThemeProp } from '../../utils';

export type ComponentSize = 'XS' | 'S' | 'M' | 'L' | 'XL';

export type ButtonType = 'primary' | 'secondary' | 'warning' | 'danger' | 'success';

export type ButtonEdges = 'squared' | 'rounded' | 'circle';

export interface Props extends Omit<HTMLProps<HTMLDivElement>, 'size' | 'type'>, WithThemeProp {
    type?: ButtonType;
    icon?: ElementType;
    textSize?: ComponentSize;
    size?: ComponentSize;
    edges?: ButtonEdges;
    fullWidth?: boolean;
    className?: string;
    children?: ReactNode;
    loading?: boolean;
    bgColor?: string;
    textColor?: string;
    hoverEffect?: boolean;
    hoverTextColor?: string;
    hoverBgColor?: string;
    hoverShadowColor?: string;
    disabled?: boolean;
}
