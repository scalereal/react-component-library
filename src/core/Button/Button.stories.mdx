import { Button, TypeMap, SizeMap } from '.';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Button"
    component={Button}
    argTypes={{
        size: {
            control: {
                type: 'select',
                options: Object.values(SizeMap)
            }
        },
        textSize: {
            control: {
                type: 'select',
                options: Object.values(SizeMap)
            }
        },
        type: {
            control: {
                type: 'select',
                options: Object.values(TypeMap)
            }
        },
        bgColor: {
            control: 'color'
        },
        textColor: {
            control: 'color'
        },
        hoverBgColor: {
            control: 'color'
        },
        hoverTextColor: {
            control: 'color'
        },
        hoverShadowColor: {
            control: 'color'
        },
        hoverEffect: {
            control: { type: 'boolean' }
        },
        disabled: {
            control: { type: 'boolean' }
        },
        ref: {
            table: {
                disable: true
            }
        },
        icon: {
            table: {
                disable: true
            }
        },
        className: {
            table: {
                disable: true
            }
        }
    }}
/>

# Button

Displays a Button

export const Template = args => (
    <div>
        <Button {...args} />
    </div>
);

<Preview>
    <Story name="Button" args={{ size: 'S', children: 'Click Me' }}>
        {Template.bind({})}
    </Story>
</Preview>

### Color Variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <div>
        <Button size="S" hoverEffect={true} bgColor="green" textColor="white" hoverBgColor="blue">
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} bgColor="red" textColor="white" hoverBgColor="green">
            Click Me
        </Button>
    </div>
</Preview>

### size

There are four sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div>
        <Button size="XS" hoverEffect={true}>
            Click Me
        </Button>
        <Button size="S" hoverEffect={true}>
            Click Me
        </Button>
        <Button size="M" hoverEffect={true}>
            Click Me
        </Button>
        <Button size="L" hoverEffect={true}>
            Click Me
        </Button>
        <Button size="XL" hoverEffect={true}>
            Click Me
        </Button>
    </div>
</Preview>

### textSize

There are four Text Sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div>
        <Button textSize="XS" hoverEffect={true}>
            Click Me
        </Button>
        <Button textSize="S" hoverEffect={true}>
            Click Me
        </Button>
        <Button textSize="M" hoverEffect={true}>
            Click Me
        </Button>
        <Button textSize="L" hoverEffect={true}>
            Click Me
        </Button>
        <Button textSize="XL" hoverEffect={true}>
            Click Me
        </Button>
    </div>
</Preview>

### Button Types

There are five types available `primary` | `secondary` | `warning` | `danger` | `success`.

<Preview withToolbar>
    <div>
        <Button size="S" hoverEffect={true} type="primary">
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} type="secondary">
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} type="warning">
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} type="danger">
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} type="success">
            Click Me
        </Button>
    </div>
</Preview>

### Disabled Button

Disabled style for all types.

<Preview withToolbar>
    <div>
        <Button size="S" hoverEffect={true} type="primary" disabled>
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} type="secondary" disabled>
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} type="warning" disabled>
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} type="danger" disabled>
            Click Me
        </Button>
        <Button size="S" hoverEffect={true} type="success" disabled>
            Click Me
        </Button>
    </div>
</Preview>

### Props

<Props story="Button" />
