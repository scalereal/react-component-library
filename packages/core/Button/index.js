var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React, { memo } from 'react';
import Text from '../Text/Text';
import { StyledButton } from './styled';
var Button = memo(React.forwardRef(function (props, ref) {
    var _a = props.type, type = _a === void 0 ? 'primary' : _a, _b = props.size, size = _b === void 0 ? 'S' : _b, textSize = props.textSize, textColor = props.textColor, hoverTextColor = props.hoverTextColor;
    return React.createElement(StyledButton, __assign({ ref: ref, type: type, size: size }, props),
        React.createElement(Text, { textSize: textSize, textColor: textColor, hoverColor: hoverTextColor }, props.children));
}));
Button.displayName = "Button";
Button.defaultProps = {
    hoverEffect: true
};
export default Button;
