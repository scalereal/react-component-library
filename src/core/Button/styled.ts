import styled from 'styled-components';
import { buttonTheme } from '../../theme/core/button';
import { defaultTheme } from '../../theme/theme';
import { ButtonEdges, ComponentSize, Props } from './types';

const SizeMixin = (size: ComponentSize) => {
    switch (size) {
        case 'XS':
            return `padding: 0.6rem 1.6rem 0.6rem;
            `;
        case 'S':
            return `padding: 0.8rem 2.4rem 0.8rem;
            `;

        case 'M':
            return `padding: 1rem 2.4rem 1rem;
           `;

        case 'L':
            return `padding: 1.2rem 2.4rem 1.2rem;
            `;

        case 'XL':
            return `padding: 1.4rem 2.4rem 1.4rem;
        `;

        default:
            return `border:1px solid red;`;
    }
};

const buttonEdges = (size: ButtonEdges) => {
    switch (size) {
        case 'squared':
            return `border-radius: 0.5rem;
            `;
        case 'rounded':
            return `border-radius: 100rem;
            `;

        case 'circle':
            return `border-radius: 50%;
           `;

        default:
            return `border-radius: 50%;`;
    }
};

const hoverPick = (pr: Props) => {
    switch (pr.type) {
        case 'primary': {
            return `background: ${pr.bgColor || buttonTheme.primary.bgColor};
            border:none;
            span{color:${pr.textColor || buttonTheme.defaults.textColor};};
            ${
                pr.hoverEffect
                    ? `&:hover {
                background-color: ${pr.hoverBgColor || buttonTheme.primary.hoverBgColor};
                box-shadow: 0px 5px 8px ${pr.hoverShadowColor || buttonTheme.defaults.hoverShadowColor};
                span{color:${pr.hoverTextColor || defaultTheme.whiteColor};};
                };`
                    : ''
            }
            `;
        }

        case 'secondary':
            return `background:transparent;
            border:1px solid ${defaultTheme.secondaryColor};
            span{color:${pr.textColor || defaultTheme.secondaryActiveColor};};
            ${
                pr.hoverEffect
                    ? `&:hover {
            border:1px solid ${defaultTheme.secondaryActiveColor};
            box-shadow: 0px 5px 8px ${pr.hoverShadowColor || defaultTheme.shadowColor};
            span{color:${pr.hoverTextColor || defaultTheme.secondaryActiveColor};};
            };`
                    : ''
            }
            `;

        case 'warning':
            return `
            background: ${pr.bgColor || defaultTheme.warningColor};
            border:none;
            span{color:${pr.textColor || defaultTheme.whiteColor};};
            ${
                pr.hoverEffect
                    ? `&:hover {
                background-color: ${pr.hoverBgColor || defaultTheme.warningActiveColor};
                box-shadow: 0px 5px 8px ${pr.hoverShadowColor || defaultTheme.shadowColor};
                span{color:${pr.hoverTextColor || defaultTheme.whiteColor};};
                };`
                    : ''
            }
           `;

        case 'danger':
            return `background: ${pr.bgColor || defaultTheme.dangerColor};
            border:none;
            span{color:${pr.textColor || defaultTheme.whiteColor};}
            ${
                pr.hoverEffect
                    ? `&:hover {
                background-color: ${pr.hoverBgColor || defaultTheme.dangerActiveColor};
                box-shadow: 0px 5px 8px ${pr.hoverShadowColor || defaultTheme.shadowColor};
                span{color:${pr.hoverTextColor || defaultTheme.whiteColor};};
                };`
                    : ''
            }
            `;

        case 'success':
            return `background: ${pr.bgColor || defaultTheme.successColor};
        border:none;
        span{color:${pr.textColor || defaultTheme.whiteColor};};
        ${
            pr.hoverEffect
                ? `&:hover {
            background-color: ${pr.hoverBgColor || defaultTheme.successActiveColor};
            box-shadow: 0px 5px 8px ${pr.hoverShadowColor || defaultTheme.shadowColor};
            span{color:${pr.hoverTextColor || defaultTheme.whiteColor};};
            };`
                : ''
        }
        `;

        default:
            return `background: ${pr.bgColor || defaultTheme.primaryColor};
        border:none;
        span{color:${pr.textColor || defaultTheme.whiteColor};};
        ${
            pr.hoverEffect
                ? `&:hover {
            background-color: ${pr.hoverBgColor || defaultTheme.primaryActiveColor};
            box-shadow: 0px 5px 8px ${pr.hoverShadowColor || defaultTheme.shadowColor};
            span{color:${pr.hoverTextColor || defaultTheme.whiteColor};};
            };`
                : ''
        }`;
    }
};

export const StyledButton = styled.button<Props>`
    ${pr => (pr.disabled ? 'pointer-events: none;opacity: 0.4;' : '')};
    ${pr => pr.fullWidth && 'width:100%;'}
    ${pr => pr.size && SizeMixin(pr.size)};
    ${pr => hoverPick(pr)};
    margin: 0.5rem;
    ${pr => pr.edges && buttonEdges(pr.edges)}
    text-align: center;
    position: relative;
    user-select: none;
    font-family: inherit;
    text-decoration: none;
    transition: all 100ms ease-out 0s;
    display: inline-flex;
    -webkit-box-pack: center;
    justify-content: center;
    -webkit-box-align: center;
    align-items: center;
    span {
        padding: 0;
    }
`;
