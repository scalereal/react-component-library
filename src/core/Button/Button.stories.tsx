import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import styled from 'styled-components';
import Button, { ButtonProps } from '.';

// import DownloadIcon from '../../icons/Download';
const ButtonType={defualt:'default', danger:'danger' , ghost:'ghost' , secondary:'secondary'};
const ButtonSize={defualt:'default', large:'large' , small:'small'};

export default {
    title: 'Components/Core/Button',
    component: Button,
    argTypes : {
        type:{
        control: {
            type: 'select',
            options: Object.keys(ButtonType),
          },
        },
    backgroundColor: { control: 'color' },
    size : {
        control: {
            type: 'select',
            options: Object.keys(ButtonSize),
          },
    },
    disabled :{
        control: {
            type: 'boolean'
        }
    },

        // children: 'Default Button',
    },
} as Meta;

const Template: Story<ButtonProps> = ({ disabled, size, backgroundColor, type, ...args }) => {
    // retrieves the appropriate icon passes it as a component prop
    const selectedType = ButtonType[type];
    const Color=backgroundColor;
    const Size = ButtonSize[size];
    const Disable = disabled;
    return <Button disabled={Disable} size={Size} backgroundColor={Color} type={selectedType} {...args} />;
  };
// const Template: Story<ButtonProps> = args => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
    children: 'Default Button',
};

// export const DangerButton = Template.bind({});
// DangerButton.args = {
//     children: 'Danger Button',
//     type: 'danger'
// };

// export const Ghost = Template.bind({});
// Ghost.args = {
//     children: 'Ghost Button',
//     type: 'ghost'
// };

// export const Secondary = Template.bind({});
// Secondary.args = {
//     children: 'Secondary Button',
//     type: 'secondary'
// };

// export const Disabled = Template.bind({});
// Disabled.args = {
//     children: 'Disabled Button',
//     disabled: true
// };

// export const Loading = Template.bind({});
// Loading.args = {
//     children: 'Button',
//     loading: true
// };

// export const WithIcon = Template.bind({});
// WithIcon.args = {
//     //   icon: DownloadIcon,
//     children: 'Download'
// };

const ButtonRow = styled.div`
    display: flex;
    align-items: flex-start;
    margin-bottom: 10px;
    & > * {
        margin-right: 10px;
    }
`;

export const Sizes = () => {
    return (
        <>
            <ButtonRow>
                <Button size="large">Large</Button>
                <Button size="default">Default</Button>
                <Button size="small">Small</Button>
            </ButtonRow>
            <ButtonRow>
                <Button type="danger" size="large">
                    Large
                </Button>
                <Button type="danger" size="default">
                    Default
                </Button>
                <Button type="danger" size="small">
                    Small
                </Button>
            </ButtonRow>
            <ButtonRow>
                <Button type="ghost" size="large">
                    Large
                </Button>
                <Button type="ghost" size="default">
                    Default
                </Button>
                <Button type="ghost" size="small">
                    Small
                </Button>
            </ButtonRow>
            <ButtonRow>
                <Button type="secondary" size="large">
                    Large
                </Button>
                <Button type="secondary" size="default">
                    Default
                </Button>
                <Button type="secondary" size="small">
                    Small
                </Button>
            </ButtonRow>
            <ButtonRow>
                <Button loading disabled size="large">
                    Large
                </Button>
                <Button loading disabled size="default">
                    Default
                </Button>
                <Button loading disabled size="small">
                    Small
                </Button>
            </ButtonRow>
        </>
    );
};
