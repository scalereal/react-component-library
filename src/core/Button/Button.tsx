import React, { FC, memo } from 'react';
import { WithStyle } from '../../utils';
import { StyledButton } from './styled';
import { Props } from './types';

export type Ref = HTMLButtonElement;

const Button:FC<Props> & WithStyle = memo(React.forwardRef( (props, ref) => {
    const {type='primary',size='S'} = props;

    return <StyledButton ref={ref} type={type} size={size} {...props}>

            {props.children}
        </StyledButton>;
}));

Button.displayName="Button";
Button.defaultProps = {
    hoverEffect: true
};

export default Button;
