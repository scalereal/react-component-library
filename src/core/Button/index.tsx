import * as React from 'react';
import { WithStyle } from '../../utils';
import { Text } from '../Text/Text';
import { StyledButton } from './styled';
import { Props } from './types';

export const SizeMap: Props['size'][] = ['XS', 'S', 'M', 'L', 'XL'];
export const TypeMap: Props['type'][] = ['primary', 'secondary', 'warning', 'danger', 'success'];
export const EdgeMap: Props['edges'][] = ['squared', 'rounded', 'circle'];

export const Button: React.FC<Props> & WithStyle = React.memo(
    React.forwardRef((props, ref) => {
        const { type = 'primary', size = 'S', textSize, textColor, hoverTextColor } = props;

        return (
            <StyledButton ref={ref} type={type} size={size} {...props}>
                <Text textSize={textSize} textColor={textColor} hoverColor={hoverTextColor}>
                    {props.children}
                </Text>
            </StyledButton>
        );
    })
);

Button.displayName = 'Button';
Button.defaultProps = {
    hoverEffect: true,
    size: 'S',
    edges: 'squared'
};

export default Button;
