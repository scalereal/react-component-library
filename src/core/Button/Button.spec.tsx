import { render } from '@testing-library/react';
import * as React from 'react';
import Button from '.';

describe('Button', () => {
    it('should render properly', () => {
        const { container } = render(<Button />);
        expect(container).toMatchSnapshot();
    });

    it('should render with all the props given', () => {
        const { container } = render(
            <Button
                size="M"
                type='danger'
                textColor='white'
                bgColor= 'red'
                hoverEffect={true}
                hoverBgColor='#126AFA'
                hoverTextColor= '#ccd2db'
                hoverShadowColor="rgba(96, 120, 144, 0.35)"
                disabled={false}
            >Button</Button>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with not-allowed cursor when it is disabled', () => {
        const { container } = render(
            <Button size="M" disabled={true} >ClickMe</Button>
        );
        expect(container).toMatchSnapshot();
    });

    // it('should render with correct children', () => {
    //     const{container}=render(<Button>Click Me</Button>);
    //     expect(container).toContainHTML("Click Me");
    // });

});