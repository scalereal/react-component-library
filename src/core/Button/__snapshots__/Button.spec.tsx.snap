// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render pro 1`] = `
<div>
  <button
    class="sc-bdfBwQ dIndFy"
    type="button"
  />
</div>
`;

exports[`Button should render properl 1`] = `
<div>
  <button
    class="sc-bdfBwQ dIndFy"
    type="button"
  />
</div>
`;

exports[`Button should render properly 1`] = `
<div>
  <button
    class="sc-bdfBwQ dIndFy"
    type="button"
  />
</div>
`;
